python/pyarrow/_fs.pyx

```diff
@@ -0,0 +1,515 @@
+    directories.
+
+    Parameters
+    ----------
+    base_dir : Union[str, pathlib.Path]
+        The directory in which to select files.
+        If the path exists but doesn't point to a directory, this should be an
+        error.
+    allow_non_existent : bool, default False
+        The behavior if `base_dir` doesn't exist in the filesystem.
+        If false, an error is returned.
+        If true, an empty selection is returned.
+    recursive : bool, default False
+        Whether to recurse into subdirectories.
+    """
+    cdef CSelector selector
+
+    def __init__(self, base_dir='', bint allow_non_existent=False,
+                 bint recursive=False):
+        self.base_dir = base_dir
+        self.recursive = recursive
+        self.allow_non_existent = allow_non_existent
+
+    @property
+    def base_dir(self):
+        return pathlib.Path(self.selector.base_dir)
+
+    @base_dir.setter
+    def base_dir(self, base_dir):
+        self.selector.base_dir = _path_to_bytes(base_dir)
+
+    @property
+    def allow_non_existent(self):
+        return self.selector.allow_non_existent
+
+    @allow_non_existent.setter
+    def allow_non_existent(self, bint allow_non_existent):
+        self.selector.allow_non_existent = allow_non_existent
+
+    @property
+    def recursive(self):
+        return self.selector.recursive
+
+    @recursive.setter
+    def recursive(self, bint recursive):
+        self.selector.recursive = recursive
+
+
+cdef class FileSystem:
+    """Abstract file system API"""
+
+    cdef:
+        shared_ptr[CFileSystem] wrapped
+        CFileSystem* fs
+
+    def __init__(self):
+        raise TypeError('dont initialize me')
+
+    cdef init(self, const shared_ptr[CFileSystem]& wrapped):
+        self.wrapped = wrapped
+        self.fs = wrapped.get()
+
+    def get_target_stats(self, paths_or_selector):
+        """Get statistics for the given target.
+
+        Any symlink is automatically dereferenced, recursively. A non-existing
+        or unreachable file returns an Ok status and has a FileType of value
+        NonExistent. An error status indicates a truly exceptional condition
+        (low-level I/O error, etc.).
+
+        Parameters
+        ----------
+        paths_or_selector: Union[Selector, List[Union[str, pathlib.Path]]]
+            Either a Selector object or a list of path-like objects.
+            The selector's base directory will not be part of the results, even
+            if it exists. If it doesn't exist, use `allow_non_existent`.
+
+        Returns
+        -------
+        file_stats : List[FileStats]
+        """
+        cdef:
+            vector[CFileStats] stats
+            vector[c_string] paths
+            CSelector selector
+
+        if isinstance(paths_or_selector, Selector):
+            selector = (<Selector>paths_or_selector).selector
+            check_status(self.fs.GetTargetStats(selector, &stats))
```

Comment:
You really should release the GIL here, unless there is a reason it doesn't work.